### PR TITLE
RNN-T hyperparamters

### PR DESCRIPTION
**Pull request by mwawrzos**

### CODE REVIEW

**Comment by johntran-nv** [comment]
@petermattson  could you please review or delegate?

**Comment by petermattson** [comment]
+Elias Mizan <emizan@google.com> could you please take a look?

**Comment by emizan76** [comment]
The only question I have is whether there is a document that justifies why some of these hyperparams are constrained. Or these constraints are "standard" for RNN-T? For example, all the DALI hyperparams are hard coded, even though we do not really need to use DALI. Does this prohibit anyone who uses DALI to change these hyperparams? Another example is opt_lamb_epsilon which can only be 1e-9, and opt_weight_decay (1e-3) while other lamb params are unconstrained.

File: training_rules.adoc [hunk 1]
```diff
@@ -332,6 +332,34 @@ The MLPerf verifier scripts checks all hyperparameters except those with names m
...
+ |rnnt |lamb |opt_lamb_beta_2                         |unconstrained |TODO |
+ |rnnt |lamb |opt_gradient_clip_norm                  |None          |TODO |
+ |rnnt |lamb |opt_gradient_accumulation_steps         |unconstrained |TODO |
+ |rnnt |lamb |opt_learning_rate_alt_decay_func        |True          |TODO |
+ |rnnt |lamb |opt_learning_rate_alt_warmup_func       |True          |TODO |
+ |rnnt |lamb |opt_lamb_learning_rate_min              |1e-5          |TODO |
+ |rnnt |lamb |train_samples                           |unconstrained |TODO |
+ |rnnt |lamb |eval_samples                            |2703          |TODO |
+ |rnnt |lamb |data_train_max_duration                 |unconstrained |TODO |
+ |rnnt |lamb |data_train_num_buckets                  |6             |TODO |
```
Comment (emizan76) [inline training_rules.adoc]: Why are most data_ hyper params constrained ?

File: training_rules.adoc [hunk 2]
```diff
@@ -332,6 +332,34 @@ The MLPerf verifier scripts checks all hyperparameters except those with names m
  |resnet |sgd |label smoothing$$*$$$$*$$ |0 or 0.1 | TODO |TODO
  |resnet |sgd |truncated norm initialization$$*$$$$*$$ |boolean | TODO |TODO
  |resnet |lars/sgd |opt_name |"lars" or "sgd" |The optimizer that was used. |
+ |rnnt |lamb |global_batch_size                       |unconstrained |reference --batch_size       |
+ |rnnt |lamb |opt_name                                |"lamb"        |The optimizer that was used. |
+ |rnnt |lamb |opt_base_learning_rate                  |unconstrained |TODO |
+ |rnnt |lamb |opt_lamb_epsilon                        |1e-9          |TODO |
```
Comment (emizan76) [inline training_rules.adoc]: Can you comment on why some lamb hyper params have hard values, while others are unconstrained ?

Comment (mwawrzos) [reply training_rules.adoc]: Beta parameters are unconstrained because in the BERT benchmark (which also uses LAMB optimizer) modifying beta parameters was necessary to converge with a higher global batch size. I don't have stable results to support the need for unfreezing it in RNN-T yet, but unfreezing seems reasonable to me.

The `opt_lamb_learning_rate_poly_decay_power` is unconstrained because of a need for alignment to a different number of steps in epoch, as the global batch size increases. The same case is for the `model_eval_ema_factor`.

I propose to unlock `opt_learning_rate_warmup_epochs` to enable extending warmup time as the learning rate increases with increased global batch size. I observed in my experiments a better convergence with the extended warmup from the baseline `6` epochs to `8`-`12` epochs with 4k GBS.

The exponential learning rate begins after `N` warmup epochs + `M` learning rate hold epochs. Increasing the number of warmups epochs helps to converge, so I proposed to unlock the `opt_lamb_learning_rate_hold_epochs`, so the learning rate can start decay in roughly the same epoch. However, I don't have enough data to prove it is necessary.

The `opt_lamb_epsilon`, because I don't expect a significant impact of this parameter on convergence.

File: training_rules.adoc (hunk 1 above)
Comment (mwawrzos) [reply training_rules.adoc]: The idea was to unfreeze parameters that don't change the training behavior significantly compared to single-node training. My understanding is, that only parameters that compensate the reduced number of mini-batches per epochs should be unfrozen, and the data pipeline hyperparameters are not in this category.

**Review by mwawrzos — COMMENTED** [review]
> The only question I have is whether there is a document that justifies why some of these hyperparams are constrained. Or these constraints are "standard" for RNN-T? For example, all the DALI hyperparams are hard coded, even though we do not really need to use DALI. Does this prohibit anyone who uses DALI to change these hyperparams? Another example is opt_lamb_epsilon which can only be 1e-9, and opt_weight_decay (1e-3) while other lamb params are unconstrained.

The `data_` hyperparameters are not precisely for DALI. The submitter can use its own methods for data preprocessing, but it should be done with the same methods as the reference data pipeline, and these methods should use the hyperparameter values that are presented here.

The epsilon and weight decay parameters are constrained because I don't expect these needs to be tuned.

The `data_train_max_duration` and `train_samples` are unconstrained due to [6.3. Data Representation](https://github.com/mlcommons/training_policies/blob/master/training_rules.adoc#63-data-representation) rules.

Reasons for unfreezing the other parameters are in other comments.

**Comment by emizan76** [comment]
Sounds good, so the following params are constrained:

All data_ parameters -- I looked at the rules and it makes sense, likewise for eval_samples

opt_lamb_epsilon: 1e-9
opt_weight_decay: 1e-3
opt_gradient_clip_norm: None
opt_learning_rate_alt_delay_func: True
opt_learning_rate_alt_warmup_func: True
opt_lamb_learning_rate_min: 1e-5

I do not see a reason for the above params being constrained, however I am OK with these hard values.
The rest are unconstrained. I will send an e-mail to the group and to see if anyone has an objection, if there is no objection after a couple of days, I will approve it so we can move on.

**Comment by emizan76** [comment]
Pasting George's and Marek's e-mail comments here, and approving since nobody has raised strong objections:

George:

> These look mostly fine to me as well. Only minor comment similar to Elias is that I don’t understand why
> lamb.opt_weight_decay is constrained to 1e-3 while opt_base_learning_rate, beta_1, beta_2 etc. are unconstrained. I can
> understand opt_lamb_epsilon being constrained cause it may not be a useful tuning parameter but opt_weight_decay could
> possibly be.

Marek

> Beta_1 and beta_2 were helpful in the case of the BERT benchmark, and I expect the same result here.
> I propose to freeze the opt_weight_decay, because it's desirable to constrain as many hyperparameters as possible to reduce > the burden on submitters to shmoo, and nobody has shown that opt_weight_decay must be unconstrained for a good
> reason.
> Can you share some reasoning for unfreezing more parameters? Maybe some data from training, or an example of
> convergence improvement/stabilization in a case of a different model, when the particular parameter is tuned.
> Unfreezing is fine if it makes the benchmark better.

George
> Here is a very recent example of BERT training where weight_decay is considered a tuning parameter:
> https://arxiv.org/pdf/1904.00962.pdf

> I agree that for the RNN-t model, we don’t have evidence whether opt_weight_decay tuning is helpful or not. In the absence
> of data going either way, I feel like that its more fair to keep it unconstrained. In the reference submission, there must be a 
> reason why opt_weight_decay is 1e-3 and not the default value of zero.

> Anyway, I don’t want to stop approval of this or anything – but probably the whole group should weigh in.

**Review by emizan76 — APPROVED** [review]
